### PR TITLE
API: Settings: delete keys that are set to null

### LINKED ISSUE #5805: WSL Integrations cannot be disabled from UI
### Actual Behavior

It is not possible to remove an existing WSL integration; while the preferences dialog accepts it, it does nothing (and reopening the dialog shows it as still checked).

### Steps to Reproduce

1. Have at least one other WSL 2 distribution.  We'll call it "OpenSUSE" below.
2. Open Preferences → WSL → Integrations.
3. Check the checkbox for OpenSUSE.
4. Accept the dialog (by hitting the Apply button).
5. Notice that Rancher Desktop does not restart the backend.
6. Open the preferences dialog again, → WSL → Integrations.
7. Notice that the checkbox is checked again.

### Result

The integration is not disabled.

### Expected Behavior

The integration should be disabled.

### Additional Information

Manually doing a `rdctl api /v1/settings --body { WSL: { integrations: { openSUSE: false } } }` does work.

### Rancher Desktop Version

1.10.0-480-g2d918f9e6

### Rancher Desktop K8s Version

N/A

### Which container engine are you using?

containerd (nerdctl)

### What operating system are you using?

Windows

### Operating System / Build Version

Windows 10 Pro 22H2 (Build 19045.3570)

### What CPU architecture are you using?

x64

### Linux only: what package format did you use to install Rancher Desktop?

None

### Windows User Only

None.

### PR DESCRIPTION
**Pull request by mook-as**
This fixes issues where WSL integrations could not be disabled.

Fixes #5805

### CODE REVIEW

File: pkg/rancher-desktop/store/preferences.ts [hunk 1]
```diff
@@ -42,31 +42,33 @@ const settingsUri = (port: number) => uri(port, 'settings');
 const lockedUri = (port: number) => uri(port, 'settings/locked');
 
 /**
- * Creates an object composed of active WSL Integrations.
- * @param integrations: The source collection, containing all WSL integrations.
- * @returns Returns a new object, containing only active WSL Integrations.
+ * Normalize WSL integrations configuration.
+ * @param integrations The source collection, containing all WSL integrations.
+ * @param mode How normalization should take place
+ * @returns Returns a new object with normalized WSL configuration.
```
Comment (ericpromislow) [inline pkg/rancher-desktop/store/preferences.ts:48]: Could you defined what normalization actually means?

For example, the `v || null` in the `submit` function maps `false` to `null`, and I don't understand why you want this.

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts [hunk 2]
```diff
@@ -511,10 +511,10 @@ export default class SettingsValidator {
       return false;
     }
 
-    let changed = false;
+    let changed = Object.keys(currentValue).some(k => !(k in desiredValue));
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/commandServer/settingsValidator.ts:514]: If you're setting `changed` here couldn't it be a constant, and we don't need the `changed ||= ...` bit at line 520 ?

**Review by ericpromislow — CHANGES_REQUESTED** [review]
When I disable an integration, it's completely removed from the settings `{... WSL: { integrations: {...}}}` instead of staying there but set to `false`.

If I wanted to manage my integrations from rdctl or the API, it's harder, because I no longer see existing but disabled integrations. I don't see why this would be a requirement.

Also a couple of comments about the code...

File: pkg/rancher-desktop/store/preferences.ts [hunk 3]
```diff
@@ -42,31 +42,33 @@ const settingsUri = (port: number) => uri(port, 'settings');
...
-    Object
-      .entries(integrations)
-      .filter(([_key, val]) => val === true),
-  );
+const normalizeWslIntegrations = (integrations: Record<string, boolean>, mode: 'diff' | 'submit') => {
+  const normalizeFn = {
+    diff:   (entries: [string, boolean][]) => entries.filter(([, v]) => v === true),
+    submit: (entries: [string, boolean][]) => entries.map(([k, v]) => [k, v || null] as const),
+  }[mode];
+  const someVal = Object.fromEntries(normalizeFn(Object.entries(integrations)));
```
Comment (ericpromislow) [inline pkg/rancher-desktop/store/preferences.ts:55]: Can just return the result of `Object.fromEntries(...)` (Same comment for the code that this PR replaces)

File: pkg/rancher-desktop/store/preferences.ts [hunk 4]
```diff
@@ -42,31 +42,33 @@ const settingsUri = (port: number) => uri(port, 'settings');
...
  */
-const pickWslIntegrations = (integrations: Record<string, boolean>) => {
-  const someVal = Object.fromEntries(
-    Object
-      .entries(integrations)
-      .filter(([_key, val]) => val === true),
-  );
+const normalizeWslIntegrations = (integrations: Record<string, boolean>, mode: 'diff' | 'submit') => {
+  const normalizeFn = {
+    diff:   (entries: [string, boolean][]) => entries.filter(([, v]) => v === true),
```
Comment (ericpromislow) [inline pkg/rancher-desktop/store/preferences.ts:52]: If `v` is boolean, can we not just say `…(([, v]) => v),` ?

File: pkg/rancher-desktop/main/commandServer/settingsValidator.ts (hunk 2 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/commandServer/settingsValidator.ts:514]: This only covers where the current value is a superset of desired value; the other part covers where they both have a key, but it's different.

File: pkg/rancher-desktop/store/preferences.ts (hunk 1 above)
Comment (mook-as) [reply pkg/rancher-desktop/store/preferences.ts:50]: Added a comment.

**Review by mook-as — COMMENTED** [review]
We remove unset integrations because otherwise it will just keep growing. You can always get a list of WSL distributions from `wsl --list`.  There should be no existing-but-disabled integrations (that's why we're doing all the syncing)…

File: pkg/rancher-desktop/store/preferences.ts (hunk 4 above)
Comment (mook-as) [reply pkg/rancher-desktop/store/preferences.ts]: Yep, was just keeping what the old code did. Changed.

File: pkg/rancher-desktop/store/preferences.ts (hunk 3 above)
Comment (mook-as) [reply pkg/rancher-desktop/store/preferences.ts]: Yep, totally doable.

**Review by ericpromislow — APPROVED** [review]
It works. Mark explained that he sets the unintegrated distros to undefined and drops them so we don't end up carrying a bunch of deleted distros in the prefs. They existing ones get read in each time anyway.

**Pull request merged**